client/src/main/java/io/getstream/chat/android/client/NewChatClientImpl.kt

```diff
@@ -0,0 +1,44 @@
+package io.getstream.chat.android.client
+
+import io.getstream.chat.android.client.api.ChatApi
+import io.getstream.chat.android.client.call.Call
+import io.getstream.chat.android.client.models.Message
+import io.getstream.chat.android.client.models.User
+import io.getstream.chat.android.client.socket.InitConnectionListener
+import io.getstream.chat.android.client.token.TokenProvider
+
+private const val NO_STORAGE_MESSAGE = "No storage was provided. If you want to have storage, provide a storageClient!"
+
+/*
+* That's the only implementation of Chat. Is shouldn't contain any business logic, this class should follow the Delegation patter.
+* The only responsibility of this class is to pass all data to the correct object to handle a specific request, nothing more.
+*
+* This class only defines a contract with the classes inside it, but the actual implementation lives outside of it.
+*
+* If some implementation is not available for this client, than no delegation is done, this way new behaviour can be created
+* by passing different implementations which conforms to the Open Close Principle. This avoid duplication of code and it uses
+* composition over inheritance.
+*/
+
+class NewChatClientImpl(
+    private val api: ChatApi,
+    private val userInitiallizer: UserInitializer,
+    private val storageClient: StorageClient?
```

Comment:
What if create two different implementation? One with storageClient as non null Type that implements `getStorageClient ` method and the other implementation throwing exception when client invokes `getStorageClient `

Reply:
For this example, that would work, but outside this POC we would have many cases that would need or not handlers. So we would need a different implementation to handle navigation, one for markdown, perhaps one for links... That would grow very fast and we would have code duplication again.